File: examples/python-sdk/test_trailing.py
#!/usr/bin/env python3
"""Exercise the Python Trailing SDK against a running server."""

from __future__ import annotations

import json
import os
import uuid

from trailing import TrailingClient


FRAMEWORK = os.getenv("TRAILING_FRAMEWORK", "eu-ai-act")


def main() -> None:
    session_id = f"python-sdk-test-{uuid.uuid4()}"
    otlp_session_id = f"{session_id}-otlp"

    with TrailingClient() as client:
        ingest_response = client.ingest(
            agent_id="python-sdk-agent",
            agent_type="gpt",
            session_id=session_id,
            action_type="tool_call",
            tool_name="query_ehr",
            target="ehr://patients/patient-1001",
            params={"patient_id": "patient-1001", "fields": ["allergies", "labs"]},
            result={"records_returned": 2, "abnormal_lab": True},
            context={"workflow": "clinical-triage", "department": "care-management"},
        )
        action_id = ingest_response["action_ids"][0]

        otlp_response = client.ingest_otel(
            {
                "resourceSpans": [
                    {
                        "scopeSpans": [
                            {
                                "spans": [
                                    {
                                        "traceId": f"trace-{uuid.uuid4().hex[:12]}",
                                        "spanId": "span-001",
                                        "name": "llm.decision",
                                        "timestamp": "2026-03-30T12:00:00Z",
                                        "attributes": {
                                            "session.id": otlp_session_id,
                                            "agent.id": "python-sdk-agent",
                                            "agent.type": "gpt",
                                            "tool.name": "risk_classifier",
                                            "target": "risk://patient-1001",
                                        },
                                        "status": "ok",
                                    }
                                ]
                            }
                        ]
                    }
                ]
            }
        )
        client.log_oversight(
            event_type="approval",
            approver="nurse.supervisor",
            scope="care-plan-release",
            related_action_id=action_id,
        )

        actions = client.get_actions(session_id=session_id)
        compliance = client.get_compliance(FRAMEWORK)
        integrity = client.verify_integrity()
        export_json = client.export_json(FRAMEWORK)
        export_pdf = client.export_pdf(FRAMEWORK)

    assert ingest_response["ingested"] == 1
    assert otlp_response["ingested"] == 1
    assert any(action["session_id"] == session_id for action in actions)
    assert compliance["framework"] == FRAMEWORK
    assert isinstance(integrity["valid"], bool)
    assert export_json["framework"] == FRAMEWORK
    assert export_pdf.startswith(b"%PDF-")

    print(
        json.dumps(
            {
                "status": "ok",
                "session_id": session_id,
                "framework": FRAMEWORK,
                "actions_seen": len(actions),
                "integrity_valid": integrity["valid"],
                "pdf_bytes": len(export_pdf),
            },
            indent=2,
        )
    )


if __name__ == "__main__":
    main()

File: examples/node-sdk/test_trailing.mjs
#!/usr/bin/env node

import assert from "node:assert/strict";
import { randomUUID } from "node:crypto";

import { TrailingClient } from "./trailing.js";

const framework = process.env.TRAILING_FRAMEWORK || "eu-ai-act";
const sessionId = `node-sdk-test-${randomUUID()}`;

const client = new TrailingClient();
const ingestResponse = await client.ingest(
  "node-sdk-agent",
  "claude",
  sessionId,
  "tool_call",
  "query_prior_auth",
  "payer://prior-auth/patient-1002",
  { patient_id: "patient-1002", service_line: "home-health" },
  { status: "expired", authorization_id: "auth-2002" },
  { workflow: "utilization-review", department: "payer-ops" },
);
const actionId = ingestResponse.action_ids[0];

const otlpResponse = await client.ingest_otel({
  resourceSpans: [
    {
      scopeSpans: [
        {
          spans: [
            {
              traceId: `trace-${randomUUID().replaceAll("-", "").slice(0, 12)}`,
              spanId: "span-001",
              name: "llm.decision",
              timestamp: "2026-03-30T12:10:00Z",
              attributes: {
                "session.id": `${sessionId}-otlp`,
                "agent.id": "node-sdk-agent",
                "agent.type": "claude",
                "tool.name": "coverage_classifier",
                target: "coverage://patient-1002",
              },
              status: "ok",
            },
          ],
        },
      ],
    },
  ],
});

const oversightResponse = await client.log_oversight(
  "override",
  "utilization.manager",
  "authorization-disposition",
  actionId,
);
const actions = await client.get_actions(sessionId);
const compliance = await client.get_compliance(framework);
const integrity = await client.verify_integrity();
const exported = await client.export_json(framework);
const pdfBytes = await client.export_pdf(framework);

assert.equal(ingestResponse.ingested, 1);
assert.equal(otlpResponse.ingested, 1);
assert.ok(oversightResponse.id);
assert.ok(actions.some((action) => action.session_id === sessionId));
assert.equal(compliance.framework, framework);
assert.equal(typeof integrity.valid, "boolean");
assert.equal(exported.framework, framework);
assert.ok(pdfBytes.subarray(0, 5).toString("utf8") === "%PDF-");

console.log(
  JSON.stringify(
    {
      status: "ok",
      sessionId,
      framework,
      actionsSeen: actions.length,
      integrityValid: integrity.valid,
      pdfBytes: pdfBytes.length,
    },
    null,
    2,
  ),
);
